### PR TITLE
BUG: Reference count leak in distance_pybind

### LINKED ISSUE #14382: Memory leak in `scipy.spatial.distance` for `cdist`
There is a memory leak in `cdist` in 1.7 that was not there in 1.6.3.

#### Reproducing code example:
```
from scipy.spatial.distance import cdist
import numpy as np

while True:
    cdist(np.random.rand(3, 2000), np.random.rand(5, 2000))
```

#### Scipy/Numpy/Python version information:
`1.7.0 1.21.0 sys.version_info(major=3, minor=9, micro=5, releaselevel='final', serial=0)`

### PR DESCRIPTION
**Pull request by peterbell10**
#### Reference issue
Fixes #14382

#### What does this implement/fix?
These `PyArray_FromAny` wrappers were using `reinterpret_borrow` which actually increments the refcount on the array, and in this case leaks the original reference. Instead, we need `reinterpret_steal` which just wraps the object without increasing the reference count, then the original reference is managed automatically.

### CODE REVIEW

**Comment by tupui** [comment]
Thanks for the quick fix! Can you share some quick results to confirm before we merge?

In general, can we somehow make some tests for memory leak?

**Review by tylerjereddy — APPROVED** [review]
If you'd like to add a test as @tupui suggests, something dead simple like this works locally:

```python
import sys

import scipy
from scipy.spatial.distance import cdist
import numpy as np

print("scipy.__version__:", scipy.__version__)

np.random.seed(123)
arr_1 = np.random.rand(3, 2000)
arr_2 = np.random.rand(5, 2000)

for i in range(5):
        cdist(arr_1, arr_2)
        print("refcount:", sys.getrefcount(arr_1))
```

`master` and SciPy `1.6.3` (before the rewrite) results below, respectively:

```
scipy.__version__: 1.8.0.dev0+1417.5e80b28
refcount: 4
refcount: 6
refcount: 8
refcount: 10
refcount: 12
```

```
scipy.__version__: 1.6.3
refcount: 2
refcount: 2
refcount: 2
refcount: 2
refcount: 2
```

File: scipy/spatial/tests/test_distance.py [hunk 1]
```diff
@@ -650,6 +651,29 @@ def test_striding(self):
...
+                    if metric == 'wminkowski':
+                        kwargs['w'] = 1.0 / x1.std(axis=0)
+
+                out = cdist(x1, x2, metric=metric, **kwargs)
+
+                # Check reference counts aren't messed up. If we only hold weak
+                # references, the arrays should be deallocated.
+                weak_refs = [weakref.ref(v) for v in (x1, x2, out)]
+                del x1, x2, out
+                assert all(weak_ref() is None for weak_ref in weak_refs)
```
Comment (peterbell10) [inline scipy/spatial/tests/test_distance.py:673]: This tests that the arrays are deallocated when all strong references are deleted.

Comment (tupui) [reply scipy/spatial/tests/test_distance.py:673]: Thank you for adding this. I didn't know about `weakref` 😃

**Review by tupui — APPROVED** [review]
LGTM, merging since there is also another approval (CI is unrelated and due to timeout). Thanks all!

**Comment by moi90** [comment]
Thanks for fixing this! It almost drove me crazy...

**Pull request merged**